File: src/test/java/com/koddy/server/coffeechat/presentation/HandlePendingCoffeeChatApiTest.java
package com.koddy.server.coffeechat.presentation;

import com.koddy.server.auth.exception.AuthExceptionCode;
import com.koddy.server.coffeechat.application.usecase.HandlePendingCoffeeChatUseCase;
import com.koddy.server.coffeechat.presentation.request.FinallyApprovePendingCoffeeChatRequest;
import com.koddy.server.coffeechat.presentation.request.FinallyCancelPendingCoffeeChatRequest;
import com.koddy.server.common.ApiDocsTest;
import com.koddy.server.common.fixture.StrategyFixture;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.SnippetFactory.body;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.SnippetFactory.path;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.createHttpSpecSnippets;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.failureDocsWithAccessToken;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.successDocsWithAccessToken;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.springframework.restdocs.payload.PayloadDocumentation.requestFields;
import static org.springframework.restdocs.request.RequestDocumentation.pathParameters;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@DisplayName("CoffeeChat -> HandlePendingCoffeeChatApi 테스트")
class HandlePendingCoffeeChatApiTest extends ApiDocsTest {
    @Autowired
    private HandlePendingCoffeeChatUseCase handlePendingCoffeeChatUseCase;

    private static final Long COFFEE_CHAT_ID = 1L;
    private final Mentor mentor = MENTOR_1.toDomain().apply(1L);
    private final Mentee mentee = MENTEE_1.toDomain().apply(2L);

    @Nested
    @DisplayName("Pending 상태인 커피챗에 대한 최종 취소 API [PATCH /api/coffeechats/pending/cancel/{coffeeChatId}]")
    class FinallyCancel {
        private static final String BASE_URL = "/api/coffeechats/pending/cancel/{coffeeChatId}";
        private final FinallyCancelPendingCoffeeChatRequest request = new FinallyCancelPendingCoffeeChatRequest("최종 취소..");

        @Test
        @DisplayName("멘토가 아니면 권한이 없다")
        void throwExceptionByInvalidPermission() {
            // given
            applyToken(true, mentee);

            // when - then
            failedExecute(
                    patchRequestWithAccessToken(new UrlWithVariables(BASE_URL, COFFEE_CHAT_ID), request),
                    status().isForbidden(),
                    ExceptionSpec.of(AuthExceptionCode.INVALID_PERMISSION),
                    failureDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Cancel/Failure", createHttpSpecSnippets(
                            pathParameters(
                                    path("coffeeChatId", "커피챗 ID(PK)", true)
                            ),
                            requestFields(
                                    body("cancelReason", "취소 사유", true)
                            )
                    ))
            );
        }

        @Test
        @DisplayName("멘토는 Pending 상태인 커피챗에 대해서 최종 취소한다")
        void success() {
            // given
            applyToken(true, mentor);
            doNothing()
                    .when(handlePendingCoffeeChatUseCase)
                    .finallyCancel(any());

            // when - then
            successfulExecute(
                    patchRequestWithAccessToken(new UrlWithVariables(BASE_URL, COFFEE_CHAT_ID), request),
                    status().isNoContent(),
                    successDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Cancel/Success", createHttpSpecSnippets(
                            pathParameters(
                                    path("coffeeChatId", "커피챗 ID(PK)", true)
                            ),
                            requestFields(
                                    body("cancelReason", "취소 사유", true)
                            )
                    ))
            );
        }
    }

    @Nested
    @DisplayName("Pending 상태인 커피챗에 대한 최종 수락 API [PATCH /api/coffeechats/pending/approve/{coffeeChatId}]")
    class FinallyApprove {
        private static final String BASE_URL = "/api/coffeechats/pending/approve/{coffeeChatId}";
        private final FinallyApprovePendingCoffeeChatRequest request = new FinallyApprovePendingCoffeeChatRequest(
                StrategyFixture.KAKAO_ID.getType().getEng(),
                StrategyFixture.KAKAO_ID.getValue()
        );

        @Test
        @DisplayName("멘토가 아니면 권한이 없다")
        void throwExceptionByInvalidPermission() {
            // given
            applyToken(true, mentee);

            // when - then
            failedExecute(
                    patchRequestWithAccessToken(new UrlWithVariables(BASE_URL, COFFEE_CHAT_ID), request),
                    status().isForbidden(),
                    ExceptionSpec.of(AuthExceptionCode.INVALID_PERMISSION),
                    failureDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Approve/Failure", createHttpSpecSnippets(
                            pathParameters(
                                    path("coffeeChatId", "커피챗 ID(PK)", true)
                            ),
                            requestFields(
                                    body("chatType", "멘토링 진행 방식", "- 링크 = zoom google" + ENTER + "- 메신저 = kakao line wechat", true),
                                    body("chatValue", "멘토링 진행 방식에 대한 값", "미팅 URL or 메신저 ID", true)
                            )
                    ))
            );
        }

        @Test
        @DisplayName("멘토는 Pending 상태인 커피챗에 대해서 최종 수락한다")
        void success() {
            // given
            applyToken(true, mentor);
            doNothing()
                    .when(handlePendingCoffeeChatUseCase)
                    .finallyApprove(any());

            // when - then
            successfulExecute(
                    patchRequestWithAccessToken(new UrlWithVariables(BASE_URL, COFFEE_CHAT_ID), request),
                    status().isNoContent(),
                    successDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Approve/Success", createHttpSpecSnippets(
                            pathParameters(
                                    path("coffeeChatId", "커피챗 ID(PK)", true)
                            ),
                            requestFields(
                                    body("chatType", "멘토링 진행 방식", "- 링크 = zoom google" + ENTER + "- 메신저 = kakao line wechat", true),
                                    body("chatValue", "멘토링 진행 방식에 대한 값", "미팅 URL or 메신저 ID", true)
                            )
                    ))
            );
        }
    }
}

File: src/test/java/com/koddy/server/coffeechat/presentation/HandlePendingCoffeeChatApiTest.kt
package com.koddy.server.coffeechat.presentation

import com.koddy.server.auth.exception.AuthExceptionCode
import com.koddy.server.coffeechat.application.usecase.HandlePendingCoffeeChatUseCase
import com.koddy.server.coffeechat.domain.model.Strategy
import com.koddy.server.coffeechat.presentation.request.FinallyApprovePendingCoffeeChatRequest
import com.koddy.server.coffeechat.presentation.request.FinallyCancelPendingCoffeeChatRequest
import com.koddy.server.common.ApiDocsTestKt
import com.koddy.server.common.docs.DocumentField
import com.koddy.server.common.docs.NUMBER
import com.koddy.server.common.docs.STRING
import com.ninjasquad.springmockk.MockkBean
import io.mockk.justRun
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest

@WebMvcTest(HandlePendingCoffeeChatApi::class)
@DisplayName("CoffeeChat -> HandlePendingCoffeeChatApi 테스트")
internal class HandlePendingCoffeeChatApiTest : ApiDocsTestKt() {
    @MockkBean
    private lateinit var handlePendingCoffeeChatUseCase: HandlePendingCoffeeChatUseCase

    @Nested
    @DisplayName("Pending 상태인 커피챗에 대한 최종 취소 API [PATCH /api/coffeechats/pending/cancel/{coffeeChatId}]")
    internal inner class FinallyCancel {
        private val baseUrl = "/api/coffeechats/pending/cancel/{coffeeChatId}"
        private val request = FinallyCancelPendingCoffeeChatRequest(cancelReason = "최종 취소..")

        private val pathParameters: Array<DocumentField> = arrayOf(
            "coffeeChatId" type NUMBER means "커피챗 ID(PK)",
        )
        private val requestFields: Array<DocumentField> = arrayOf(
            "cancelReason" type STRING means "최종 취소 사유",
        )

        @Test
        fun `멘토가 아니면 권한이 없다`() {
            val exceptionCode = AuthExceptionCode.INVALID_PERMISSION

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(mentee)
                bodyContent(request)
            }.andExpect {
                status { isForbidden() }
                content { exception(exceptionCode) }
            }.andDo {
                makeFailureDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Cancel/Failure") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }

        @Test
        fun `멘토는 Pending 상태인 커피챗에 대해서 최종 취소한다`() {
            justRun { handlePendingCoffeeChatUseCase.finallyCancel(any()) }

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(mentor)
                bodyContent(request)
            }.andExpect {
                status { isNoContent() }
            }.andDo {
                makeSuccessDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Cancel/Success") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }
    }

    @Nested
    @DisplayName("Pending 상태인 커피챗에 대한 최종 수락 API [PATCH /api/coffeechats/pending/approve/{coffeeChatId}]")
    internal inner class FinallyApprove {
        private val baseUrl = "/api/coffeechats/pending/approve/{coffeeChatId}"
        private val request = FinallyApprovePendingCoffeeChatRequest(
            chatType = Strategy.Type.KAKAO_ID.value,
            chatValue = "sjiwon",
        )

        private val pathParameters: Array<DocumentField> = arrayOf(
            "coffeeChatId" type NUMBER means "커피챗 ID(PK)",
        )
        private val requestFields: Array<DocumentField> = arrayOf(
            "chatType" type STRING means "멘토링 진행 방식" constraint "- 링크 = zoom, google $enter - 메신저 = kakao, line, wechat",
            "chatValue" type STRING means "멘토링 진행 방식에 대한 값" constraint "링크 or 메신저 ID",
        )

        @Test
        fun `멘토가 아니면 권한이 없다`() {
            val exceptionCode = AuthExceptionCode.INVALID_PERMISSION

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(mentee)
                bodyContent(request)
            }.andExpect {
                status { isForbidden() }
                content { exception(exceptionCode) }
            }.andDo {
                makeFailureDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Approve/Failure") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }

        @Test
        fun `멘토는 Pending 상태인 커피챗에 대해서 최종 수락한다`() {
            justRun { handlePendingCoffeeChatUseCase.finallyApprove(any()) }

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(mentor)
                bodyContent(request)
            }.andExpect {
                status { isNoContent() }
            }.andDo {
                makeSuccessDocsWithAccessToken("CoffeeChatApi/LifeCycle/PendingCoffeeChat/Approve/Success") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }
    }
}
